src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -215,13 +215,18 @@ InflightData GetInflightData (NSUrlSessionTask task)
 					if (sessionHandler.inflightRequests.TryGetValue (task, out inflight))
 						return inflight;
 
+				// if we did not manage to get the inflight data, we either got an error or have been canceled, lets cancel the task, that will execute DidCompleteWithError
+				task.Cancel ();
 				return null;
 			}
 
 			public override void DidReceiveResponse (NSUrlSession session, NSUrlSessionDataTask dataTask, NSUrlResponse response, Action<NSUrlSessionResponseDisposition> completionHandler)
 			{
 				var inflight = GetInflightData (dataTask);
 
+				if (inflight == null)
+					return;
+
 				try {
 					// ensure that we did not cancel the request, if we did, do cancel the task
 					if (inflight.CancellationToken.IsCancellationRequested) {
```

Comment:
@mandel-macaque Let's go with the first once, it's closer to the existing behaviour.